enclave-bridge/src/lib.rs

```diff
@@ -261,6 +312,49 @@ impl<T: Config> Pallet<T> {
 		)
 	}
 
+	pub fn get_maybe_updated_shard_config(
+		shard: ShardIdentifier,
+		apply_due_update: bool,
+	) -> Option<ShardConfig<T::AccountId>> {
+		let current_block_number = <frame_system::Pallet<T>>::block_number();
+		Self::shard_config(shard).map(|current| {
+			current.upgrade_at.filter(|&at| at <= current_block_number).map_or_else(
+				|| current.active_config.clone(),
+				|_| {
+					current.pending_upgrade.map_or(current.active_config.clone(), |due_update| {
+						if apply_due_update {
+							<ShardConfigRegistry<T>>::insert(
+								shard,
+								UpgradableShardConfig::from(due_update.clone()),
+							);
+						}
+						due_update
+					})
+				},
+			)
+		})
+	}
+
+	#[allow(clippy::type_complexity)]
+	pub fn get_sovereign_enclave_and_touch_shard(
+		enclave_signer: &T::AccountId,
+		shard: ShardIdentifier,
+	) -> Result<
+		(MultiEnclave<Vec<u8>>, Vec<ShardSignerStatus<T::AccountId, T::BlockNumber>>),
+		DispatchErrorWithPostInfo,
+	> {
+		let enclave = Teerex::<T>::get_sovereign_enclave(enclave_signer)?;
+		ensure!(
+			enclave.fingerprint() ==
+				Self::get_maybe_updated_shard_config(shard, true)
+					.unwrap_or_else(|| ShardConfig::new(shard))
+					.enclave_fingerprint,
+			<Error<T>>::WrongFingerprintForShard
+		);
```

Comment:
With the above mentioned suggestions about separation of concerns, we can simplify this to:

```rust

if let Some(shard_update) = Self::get_due_update(shard) {
     	<ShardConfigRegistry<T>>::insert(
		shard,
		// no update here
		ShardConfig::from(due_update.clone()),
	);

	ensure!(enclave.finger_print() == shard_update.finger_print, ...);
}

// where get due update is very simple:
	fn get_due_update(shard: ShardIdentifier) -> Option<ShardUpdate> {
		let current_block_number = <frame_system::Pallet<T>>::block_number();

		Self::pending_shard_config_update(shard)
			.filter(|pending_update| pending_update.update_at <= current_block_number)
	}

This also removes a redundant check, that the new shard fingerprint matches the enclave creating the new shard.
```